posts/2020/06/portable-python-env.md

```diff
@@ -0,0 +1,184 @@
+
+```yaml
+required:
+- numpy
+- pandas
+- scikit-learn
+- scipy
+```
+
+### Create a temporary environment
+
+Now we have a list of the direct dependencies but what about all the sub-dependencies?
+To capture these, we'll create a temporary environment.
+
+Copy the yaml formatted dependencies into an environment file named `environment.yml`.
+
+```yaml
+name: my_env
+channels:
+  - conda-forge
+dependencies:
+  - python>=3.7
+  - numpy
+  - pandas
+  - scikit-learn
+  - scipy
+  - conda-pack
+```
+
+Notice that we've added two extra packages to our `environment.yml`.
+In this example, we'll set a minimum python version to include in the package.
+We could also have explicitly set the Python version. You may notice that we
+have also added an additional package called called `conda-pack`. This will be used
+for wrapping up the environment for distribution - more on that later.
+
+Create a conda environment from this yaml that will include all of the necessary
+dependencies.
+
+`conda env create -f environment.yml`
+
+Activate the temporary conda env:
+
+`conda activate my_env`
+
+## Wrap up the environment into a tarball
+
+At this point, we're ready to wrap up our environment into a single tarball.
+To do this, we'll use a package called `conda-pack`. `Conda-pack` is going to help us
+wrap up our exact environment, including python itself. This means that the target machine
+is not required to have python installed for this environment to be utilized. Much of what
+follows is taken directly from the `conda-pack` docs.
+
+Pack environment my_env into out_name.tar.gz
+
+`conda pack -n my_env -o my_env.tar.gz`
+
+## Unpacking the environment on the target machine
+
+At this point you will have a portable tarball that you can send to a different
+machine. Note that the tarball you've created must only be used on target machines
+with the same operating system.
+
+Now we'll go over how to unpack the tarball on the target machine and utilize this
+environment.
+
+Unpack environment into directory `my_env`:
+
+`$ mkdir -p my_env`
+`$ tar -xzf my_env.tar.gz -C my_env`
+
+We could stop here and start using the python environment directly. Note that most
+Python libraries will work fine, but things that require prefix cleanups (since
+we've built it in one directory and moved to another) will fail.
+
+`$ ./my_env/bin/python`
+
+Alternatively we could activate the environment. This adds `my_env/bin` to your path
+
+`$ source my_env/bin/activate`
+
+And then run python from the activated environment
+
+`(my_env) $ python`
+
+Cleanup prefixes from inside the active environment.
+Note that this command can also be run without activating the environment
+as long as some version of python is already installed on the machine.
+
+`(my_env) $ conda-unpack`
+
+At this point the environment is exactly as if you installed it here
+using conda directly. All scripts should be fully functional, e.g.:
+
+`(my_env) $ ipython --version`
```

Comment:
Don’t know if it matters for this example, but ipython wasn’t one of the explicit dependencies. Will it be installed?

Reply:
Good catch! I added it